plugin/path_impersonated_account.go

```diff
@@ -0,0 +1,253 @@
+}
+
+func (b *backend) pathImpersonatedAccountRead(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	nameRaw, ok := d.GetOk("name")
+	if !ok {
+		return logical.ErrorResponse("name is required"), nil
+	}
+
+	acct, err := b.getImpersonatedAccount(nameRaw.(string), ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if acct == nil {
+		return nil, nil
+	}
+
+	data := map[string]interface{}{
+		"service_account_project": acct.Project,
+		"service_account_email":   acct.EmailOrId,
+		"token_scopes":            acct.TokenScopes,
+		"ttl":                     acct.Ttl,
+	}
+
+	return &logical.Response{
+		Data: data,
+	}, nil
+}
+
+func (b *backend) pathImpersonatedAccountDelete(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	nameRaw, ok := d.GetOk("name")
+	if !ok {
+		return logical.ErrorResponse("name is required"), nil
+	}
+	name := nameRaw.(string)
+
+	b.impersonatedAccountLock.Lock()
+	defer b.impersonatedAccountLock.Unlock()
+
+	// Delete impersonated account
+	b.Logger().Debug("deleting impersonated account from storage", "name", name)
+	if err := req.Storage.Delete(ctx, fmt.Sprintf("%s/%s", impersonatedAccountStoragePrefix, name)); err != nil {
+		return nil, err
+	}
+
+	b.Logger().Debug("finished deleting impersonated account from storage", "name", name)
+	return nil, nil
+}
+
+func (b *backend) pathImpersonatedAccountCreate(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	input, warnings, err := b.parseImpersonateInformation(nil, d)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+	if input == nil {
+		return nil, fmt.Errorf("plugin error - parse returned unexpected nil input")
+	}
+
+	b.impersonatedAccountLock.Lock()
+	defer b.impersonatedAccountLock.Unlock()
+
+	// Create and save impersonated account with new resources.
+	if err := b.createImpersonatedAccount(ctx, req, input); err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+	if len(warnings) > 0 {
+		return &logical.Response{Warnings: warnings}, nil
+	}
+	return nil, nil
+}
+
+func (b *backend) pathImpersonatedAccountUpdate(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	nameRaw, ok := d.GetOk("name")
+	if !ok {
+		return logical.ErrorResponse("name is required"), nil
+	}
+	name := nameRaw.(string)
+
+	b.impersonatedAccountLock.Lock()
+	defer b.impersonatedAccountLock.Unlock()
+
+	acct, err := b.getImpersonatedAccount(name, ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if acct == nil {
+		return nil, fmt.Errorf("unable to find impersonated account %s to update", name)
+	}
+
+	initialInput := &ImpersonatedAccount{
+		Name:        acct.Name,
+		Ttl:         acct.Ttl,
+		TokenScopes: acct.TokenScopes,
+		ServiceAccountId: gcputil.ServiceAccountId{
+			Project:   acct.Project,
+			EmailOrId: acct.EmailOrId,
+		},
+	}
```

Comment:
Kind of a nit, but I wonder if we could pass `acct` directly to `b.parseImpersonateInformation(acct, d)` instead of copying each of its fields into `initialInput`? Is there any value in doing this copy over to a new struct? We could delete lines 173-181 if there is no difference.

Reply:
Nice! Cleans up really nicely and both unit and integration tests still pass.